prow/cmdutil/github.go

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmdutil
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+
+	"k8s.io/test-infra/cmdutil"
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/flagutil"
+	"k8s.io/test-infra/prow/git"
+	"k8s.io/test-infra/prow/github"
+)
+
+// GitHubOptions holds options for interacting with GitHub.
+type GitHubOptions struct {
+	endpoint      flagutil.Strings
+	TokenPath     string
+	TokenRequired bool
+}
+
+// AddFlags injects GitHub options into the given FlagSet.
+func (o *GitHubOptions) AddFlags(fs *flag.FlagSet) {
+	o.endpoint = flagutil.NewStrings("https://api.github.com")
+	fs.Var(&o.endpoint, "github-endpoint", "GitHub's API endpoint (may differ for enterprise).")
+	fs.StringVar(&o.TokenPath, "github-token-path", "/etc/github/oauth", "Path to the file containing the GitHub OAuth secret.")
+	fs.StringVar(&o.TokenPath, "github-token-file", "/etc/github/oauth", "DEPRECATED: use -github-token-path instead.")
+}
+
+// Validate validates GitHub options.
+func (o *GitHubOptions) Validate(dryRun bool) error {
+	if err := cmdutil.ValidateURIs("-github-endpoint", o.endpoint.Strings()); err != nil {
+		return err
+	}
+
+	if o.TokenRequired && o.TokenPath == "" {
+		return errors.New("empty -github-token-path")
+	}
+
+	return nil
+}
+
+// GitHubClient returns a GitHub client.
+func (o *GitHubOptions) GitHubClient(secretAgent *config.SecretAgent, dryRun bool) (client *github.Client, err error) {
+	var generator *func() []byte
+	if o.TokenPath == "" {
+		generatorFunc := func() []byte {
+			return []byte{}
+		}
+		generator = &generatorFunc
+	} else {
+		if secretAgent == nil {
+			return nil, fmt.Errorf("cannot store token from %q without a secret agent", o.TokenPath)
+		}
+		generatorFunc := secretAgent.GetTokenGenerator(o.TokenPath)
+		generator = &generatorFunc
+	}
+
+	if dryRun {
+		return github.NewDryRunClient(*generator, o.endpoint.Strings()...), nil
+	}
+	return github.NewClient(*generator, o.endpoint.Strings()...), nil
+}
+
+// GitClient returns a Git client.
+func (o *GitHubOptions) GitClient(secretAgent *config.SecretAgent, dryRun bool) (client *git.Client, err error) {
```

Comment:
Why is this tied to the GitHub flags?

Reply:
> Why is this tied to the GitHub flags?

Because [we seed it with the GitHub bot name][1].

[1]: https://github.com/kubernetes/test-infra/pull/9503/commits/8e4a1c4eacf6f4e9fe3c47a5e2f75727b53ac2d7#diff-10d11420786ce54a3c335b47d1a5e063R102